src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1229,12 +1229,12 @@ public void M3()
                 // (5,23): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
                 //     public async void M()
                 Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "M").WithLocation(5, 23),
-                // (16,17): error CS0610: Field or property cannot be of type 'ArgIterator'
+                // (16,16): error CS0306: The type 'ArgIterator' may not be used as a type argument
                 //         return (default(ArgIterator), default(ArgIterator));
-                Diagnostic(ErrorCode.ERR_FieldCantBeRefAny, "default(ArgIterator)").WithArguments("System.ArgIterator").WithLocation(16, 17),
-                // (16,39): error CS0610: Field or property cannot be of type 'ArgIterator'
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "(default(ArgIterator), default(ArgIterator))").WithArguments("System.ArgIterator").WithLocation(16, 16),
+                // (16,16): error CS0306: The type 'ArgIterator' may not be used as a type argument
                 //         return (default(ArgIterator), default(ArgIterator));
-                Diagnostic(ErrorCode.ERR_FieldCantBeRefAny, "default(ArgIterator)").WithArguments("System.ArgIterator").WithLocation(16, 39)
+                Diagnostic(ErrorCode.ERR_BadTypeArgument, "(default(ArgIterator), default(ArgIterator))").WithArguments("System.ArgIterator").WithLocation(16, 16)
```

Comment:
The fact that deconstruction tests are affected is a reminder that deconstructions produce tuples as well. It looks like we added a checked there for restricted types (like ArgIterator), but we should apply the full `CheckConstraints`. 

For instance, a deconstruction that produces a tuple of pointer types should get errors, or a deconstruction that produces a tuple when ValueTuple is specially-coded to have constraints. We need to test those (the same bug probably exists in those code paths).
`MakeDeconstructionConstructionStep` and `MakeMergedTupleType` are relevant methods to investigate (they create tuples).